core/chains/evm/logpoller/orm.go

```diff
@@ -320,27 +333,46 @@ func (o *DSORM) DeleteBlocksBefore(ctx context.Context, end int64, limit int64)
 	}
 
 	// Remove up to limit blocks at a time, until we've reached the limit or removed everything eligible for deletion
-	var deleted, rows int64
-	for limitBlock += (limit - 1); deleted < limit; limitBlock += limit {
-		if limitBlock > end {
-			limitBlock = end
-		}
-		result, err = o.ds.ExecContext(ctx, `DELETE FROM evm.log_poller_blocks WHERE block_number <= $1 AND evm_chain_id = $2`, limitBlock, ubig.New(o.chainID))
-		if err != nil {
-			return deleted, err
+	var upper int64
+	for lower := start; rowsAffected < limit; lower = upper + 1 {
+		upper = lower + limit - 1
+		if upper > end {
+			upper = end
 		}
 
-		if rows, err = result.RowsAffected(); err != nil {
-			return deleted, err
+		rows, err2 := r.query(ctx, r, lower, upper)
+		if err2 != nil {
+			return rowsAffected, err2
 		}
+		rowsAffected += rows
 
-		deleted += rows
-
-		if limitBlock == end {
+		if upper == end {
```

Comment:
nit, but probably `if upper >= end { break }` would be safer. Not sure if there is a real threat of accidentally going beyond, but `upper >= end` will prevent from that

Reply:
Just a few lines before this is:
```
		if upper > end {
			upper = end
		}
```
so definitely not possible ;-) But sure... marginally safer in case someone tries to modify the code in the future without thinking carefully